internal/server/storage/postgres/postgre.go

```diff
@@ -0,0 +1,465 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"log"
+	"strconv"
+	"time"
+
+	"github.com/jackc/pgerrcode"
+	"github.com/jackc/pgx/v5/pgconn"
+
+	"github.com/Arcadian-Sky/musthave-metrics/internal/server/models"
+	"github.com/Arcadian-Sky/musthave-metrics/internal/server/storage"
+)
+
+// PostgresStorage представляет хранилище метрик в PostgreSQL
+type PostgresStorage struct {
+	db                *sql.DB
+	retriableErrorMap map[string]bool
+	maxRetries        int
+	initialDelay      time.Duration
+}
+
+// NewPostgresStorage создает новый экземпляр PostgresStorage
+func NewPostgresStorage(db *sql.DB) *PostgresStorage {
+	p := &PostgresStorage{
+		db: db,
+		retriableErrorMap: map[string]bool{
+			pgerrcode.UniqueViolation: true,
+		},
+		maxRetries:   3,
+		initialDelay: time.Second,
+	}
+	err := p.CreateMetricsTable()
+	if err != nil {
+		panic(err)
+	}
+	return p
+}
+
+// executeWithRetry - функция для выполнения операции с повторными попытками
+func (p *PostgresStorage) executeWithRetry(ctx context.Context, operation func() error) error {
+	var err error
+	delay := p.initialDelay
+	for attempt := 0; attempt <= p.maxRetries; attempt++ {
```

Comment:
SUGGESTION: if I were you, I would use the `backoff` package.
https://github.com/cenkalti/backoff